security/advancedtls/crl_test.go

```diff
@@ -731,3 +731,35 @@ func TestIssuerNonPrintableString(t *testing.T) {
 		t.Fatalf("fetchCRL failed: %s", err)
 	}
 }
+
+func TestCRLCacheExpirationReloading(t *testing.T) {
+	cache, err := lru.New(5)
+	if err != nil {
+		t.Fatalf("Creating cache failed")
+	}
+
+	var certs = makeChain(t, testdata.Path("crl/revokedInt.pem"))
+	// Certs[1] has the same issuer as the revoked cert
+	rawIssuer := certs[1].RawIssuer
+
+	// `3.crl`` revokes `revokedInt.pem`
+	crl := loadCRL(t, testdata.Path("crl/3.crl"))
+	// Modify the crl so that the cert is NOT revoked and add it to the cache
+	crl.CertList.TBSCertList.RevokedCertificates = nil
+	crl.CertList.TBSCertList.NextUpdate = time.Now().Add(time.Hour)
+	cache.Add(hex.EncodeToString(rawIssuer), crl)
+	var cfg = RevocationConfig{RootDir: testdata.Path("crl"), Cache: cache}
+	revocationStatus := checkChain(certs, cfg)
+	if revocationStatus != RevocationUnrevoked {
```

Comment:
This scenario might not happen in the real world...the reason why the certs are unrevoked is that you load the CRL without a handshake going on, so the cache won't pick up that CRL.
In reality, users set this through advancedtls APIs, so it is in use whenever a new handshake takes place. 
A better scenario might be:
loading `3.crl` into cache, and it revokes `revokedInt.pem` -> update `3.crl`'s NextUpdate, and cause the cache to load `2.crl` ->  `2.crl` should replace `3.crl` in cache and it doesn't revoke `revokedInt.pem`

Reply:
I don't believe that un-revocation is something we explicitly support (even if the code makes it viable right now), so I hesitate to design the test for that. I'd rather test the unrevoked -> revoked path.

The issue is that we want to mimic a new CRL coming into the system - I could design this even more e2e such that we have a temporary folder created just for this test, place a CRL with no revocation in it, handshake, then copy the hashed `3.crl` into the directory representing an updated CRL in which `revokedInt.pem` is revoked.